malfeasance2/handler.go

```diff
@@ -0,0 +1,140 @@
+package malfeasance2
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"go.uber.org/zap"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+var (
+	ErrMalformedData  = fmt.Errorf("%w: malformed data", pubsub.ErrValidationReject)
+	ErrWrongHash      = fmt.Errorf("%w: incorrect hash", pubsub.ErrValidationReject)
+	ErrUnknownVersion = fmt.Errorf("%w: unknown version", pubsub.ErrValidationReject)
+	ErrUnknownDomain  = fmt.Errorf("%w: unknown domain", pubsub.ErrValidationReject)
+)
+
+type Handler struct {
+	logger   *zap.Logger
+	db       sql.Executor
+	self     p2p.Peer
+	tortoise tortoise
+
+	handlers map[ProofDomain]MalfeasanceHandler
+}
+
+func NewHandler(
+	db sql.Executor,
+	lg *zap.Logger,
+	self p2p.Peer,
+	tortoise tortoise,
+) *Handler {
+	return &Handler{
+		db:       db,
+		logger:   lg,
+		self:     self,
+		tortoise: tortoise,
+
+		handlers: make(map[ProofDomain]MalfeasanceHandler),
+	}
+}
+
+func (h *Handler) RegisterHandler(malfeasanceType ProofDomain, handler MalfeasanceHandler) {
+	h.handlers[malfeasanceType] = handler
+}
+
+func (h *Handler) HandleSynced(ctx context.Context, expHash types.Hash32, _ p2p.Peer, msg []byte) error {
+	var proof MalfeasanceProof
+	if err := codec.Decode(msg, &proof); err != nil {
+		h.logger.Warn("failed to decode malfeasance proof",
+			zap.Stringer("exp_hash", expHash),
+			zap.Error(err),
+		)
+		return ErrMalformedData
+	}
+
+	nodeIDs, err := h.handleProof(ctx, proof)
+	if err != nil {
+		h.logger.Warn("failed to handle synced malfeasance proof",
+			zap.Stringer("exp_hash", expHash),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	if !slices.Contains(nodeIDs, types.NodeID(expHash)) {
+		h.logger.Warn("synced malfeasance proof invalid for requested nodeID",
+			zap.Stringer("exp_hash", expHash),
+		)
+		return ErrWrongHash
+	}
+
+	if err := h.storeProof(ctx, proof.Domain, msg); err != nil {
+		h.logger.Warn("failed to store synced malfeasance proof",
+			zap.Stringer("exp_hash", expHash),
+			zap.Error(err),
+		)
+		return err
+	}
+	return nil
+}
+
+func (h *Handler) HandleGossip(ctx context.Context, peer p2p.Peer, msg []byte) error {
```

Comment:
nit: lots of logging in this function can be avoided and in general better to avoid log-and-return pattern

Reply:
also, this method is almost an exact copy of the one above